File: mac/Sources/Crypto.swift
import Foundation
import CryptoKit

enum CryptoError: Error { case badKey, badMessage }

/// AES-256-GCM with the pre-shared pairing key. `ct` on the wire is
/// ciphertext followed by the 16-byte tag. AAD is the room string's bytes.
struct Crypto {
    private let key: SymmetricKey
    private let roomAAD: Data

    init(keyBase64: String, room: String) throws {
        guard let kd = Data(base64Encoded: keyBase64), kd.count == 32 else {
            throw CryptoError.badKey
        }
        self.key = SymmetricKey(data: kd)
        self.roomAAD = Data(room.utf8)
    }

    func open(nonceB64: String, ctB64: String) throws -> Data {
        guard let nonceData = Data(base64Encoded: nonceB64),
              let ctFull = Data(base64Encoded: ctB64),
              ctFull.count > 16 else { throw CryptoError.badMessage }
        let nonce = try AES.GCM.Nonce(data: nonceData)
        let tag = ctFull.suffix(16)
        let ct = ctFull.prefix(ctFull.count - 16)
        let box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ct, tag: tag)
        return try AES.GCM.open(box, using: key, authenticating: roomAAD)
    }

    /// Kept for tests and future Mac->phone messaging.
    func seal(_ plaintext: Data) throws -> (nonce: String, ct: String) {
        let box = try AES.GCM.seal(plaintext, using: key, authenticating: roomAAD)
        return (Data(box.nonce).base64EncodedString(),
                (box.ciphertext + box.tag).base64EncodedString())
    }

    // Raw binary variants for file-transfer frames (avoid base64 overhead).

    /// Returns nonce(12) || ciphertext || tag(16).
    func sealRaw(_ plaintext: Data) throws -> Data {
        let box = try AES.GCM.seal(plaintext, using: key, authenticating: roomAAD)
        return Data(box.nonce) + box.ciphertext + box.tag
    }

    func openRaw(nonce: Data, ctTag: Data) throws -> Data {
        guard ctTag.count >= 16 else { throw CryptoError.badMessage }
        let sealedNonce = try AES.GCM.Nonce(data: nonce)
        let tag = ctTag.suffix(16)
        let ct = ctTag.prefix(ctTag.count - 16)
        let box = try AES.GCM.SealedBox(nonce: sealedNonce, ciphertext: ct, tag: tag)
        return try AES.GCM.open(box, using: key, authenticating: roomAAD)
    }
}

File: android/app/src/main/java/com/hari/otpbridge/Crypto.kt
package com.hari.otpbridge

import android.util.Base64
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * AES-256-GCM matching PROTOCOL.md (verified byte-compatible with the Node
 * relay and the Swift Mac app). `ct` = ciphertext||tag; AAD = utf8(room).
 */
object Crypto {
    private val rng = SecureRandom()

    fun seal(keyB64: String, room: String, plaintext: String): Envelope {
        val key = SecretKeySpec(Base64.decode(keyB64, Base64.NO_WRAP), "AES")
        val nonce = ByteArray(12).also { rng.nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, key, GCMParameterSpec(128, nonce))
        cipher.updateAAD(room.toByteArray(Charsets.UTF_8))
        val ctFull = cipher.doFinal(plaintext.toByteArray(Charsets.UTF_8))
        return Envelope(
            room = room,
            nonce = Base64.encodeToString(nonce, Base64.NO_WRAP),
            ct = Base64.encodeToString(ctFull, Base64.NO_WRAP),
        )
    }

    /** Raw binary seal for file frames: returns nonce(12) || ciphertext || tag. */
    fun sealRaw(keyB64: String, room: String, plaintext: ByteArray): ByteArray {
        val key = SecretKeySpec(Base64.decode(keyB64, Base64.NO_WRAP), "AES")
        val nonce = ByteArray(12).also { rng.nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, key, GCMParameterSpec(128, nonce))
        cipher.updateAAD(room.toByteArray(Charsets.UTF_8))
        return nonce + cipher.doFinal(plaintext)   // nonce || ct || tag
    }

    /** Raw binary open for file frames. nonce = 12 bytes, ctTag = ciphertext||tag. */
    fun openRaw(keyB64: String, room: String, nonce: ByteArray, ctTag: ByteArray): ByteArray {
        val key = SecretKeySpec(Base64.decode(keyB64, Base64.NO_WRAP), "AES")
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(128, nonce))
        cipher.updateAAD(room.toByteArray(Charsets.UTF_8))
        return cipher.doFinal(ctTag)
    }

    /** Kept for future Mac->phone messages. */
    fun open(keyB64: String, room: String, nonceB64: String, ctB64: String): String {
        val key = SecretKeySpec(Base64.decode(keyB64, Base64.NO_WRAP), "AES")
        val nonce = Base64.decode(nonceB64, Base64.NO_WRAP)
        val ctFull = Base64.decode(ctB64, Base64.NO_WRAP)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(128, nonce))
        cipher.updateAAD(room.toByteArray(Charsets.UTF_8))
        return String(cipher.doFinal(ctFull), Charsets.UTF_8)
    }
}
